### PR TITLE
PromptLayer integration

### PR DESCRIPTION
**Pull request by bmax**
# Description

Integrate promptlayer.com cc @jped

## Type of Change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Added new unit/integration tests
- [x] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

# Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added Google Colab support for the newly added notebooks.
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: llama_index/callbacks/promptlayer_handler.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        if completion:
+            function_name = PROMPT_LAYER_COMPLETION_FUNCTION_NAME
+            resp = str(completion)
+        pl_request_id = self._promptlayer_api_request(
+            function_name,
+            "openai",
+            [prompt],
+            {
+                **extra_args,
+                **event_data["kwargs"],
```
Comment (bmax) [inline llama_index/callbacks/promptlayer_handler.py:109]: I think prompt layer can support token usage, got to figure out how to pass that in as well, does openai respond w/ token usage or do we need to use token_counter callback?

File: llama_index/callbacks/promptlayer_handler.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    ) -> None:
+        if event_type != CBEventType.LLM or payload is None:
+            return
+        request_end_time = datetime.datetime.now().timestamp()
+        prompt = str(payload.get(EventPayload.PROMPT))
+        completion = payload.get(EventPayload.COMPLETION)
+        response = payload.get(EventPayload.RESPONSE)
+        function_name = PROMPT_LAYER_CHAT_FUNCTION_NAME
+        event_data = self.get_event(event_id=event_id)
+        resp: str | Dict
```
Comment (bmax) [inline llama_index/callbacks/promptlayer_handler.py:85]: how should I handle this? apparently `|` is only supported in 3.10

File: llama_index/callbacks/promptlayer_handler.py (hunk 1 above)
Comment (Jped) [reply llama_index/callbacks/promptlayer_handler.py:109]: You dont not need to send token count, if you have it readily available you send send them the same way openai sends them which is 
```
"usage": {
    "prompt_tokens": 9,
    "completion_tokens": 12,
    "total_tokens": 21
  }
```
in the function kwargs

File: llama_index/callbacks/promptlayer_handler.py (hunk 2 above)
Comment (logan-markewich) [reply llama_index/callbacks/promptlayer_handler.py:85]: Use `Union[str, Dict]`

**Pull request merged**